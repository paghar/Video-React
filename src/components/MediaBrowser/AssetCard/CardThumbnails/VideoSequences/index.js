//@flow
import React, { useState, useRef, useEffect } from 'react';
import { SequenceHolder, SequencesWrapper, OffsetBar } from './index.style';

type SequenceSizeType = {
  width: number,
  height: number,
};

type Props = {
  sequencesSrc: string,
  sequencesCount: number,
  sequenceSize: SequenceSizeType,
  className?: string,
  expanded?: boolean,
};

const VideoSequences = ({
  className,
  sequencesSrc,
  sequencesCount,
  sequenceSize,
  expanded,
}: Props) => {
  const wrapperRef = useRef(null); // ref to component wrapper

  // This function calculates where shouldbe sequences image position
  const calculateSequenceOffset = currentSequence => {
    if (wrapperRef.current) {
      const {
        width: wrapperWidth,
      } = wrapperRef.current.getBoundingClientRect();
      return (
        (Math.floor(currentSequence) * sequenceSize.height) /
        (sequenceSize.width / wrapperWidth)
      );
    } else {
      return 0;
    }
  };

  const [barOffset, setBarOffset] = useState(-1); // bar x offset to be used in translateX
  const [sequencesOffset, setSequencesOffset] = useState(0); // sequences offset y to be used in backgroundPositionY

  // This function handles moving bar and changing sequence number according to mouse position to be shown
  const handleMouseMove = ({ pageX }) => {
    if (wrapperRef.current) {
      const {
        x: wrapperX,
        left: wrapperLeft,
        width: wrapperWidth,
      } = wrapperRef.current.getBoundingClientRect();

      const barOffset = pageX - (wrapperX || wrapperLeft);
      const currentSequence = (barOffset / wrapperWidth) * sequencesCount;
      const sequencesOffset = calculateSequenceOffset(currentSequence);
      setBarOffset(barOffset);
      setSequencesOffset(sequencesOffset);
    }
  };

  // This function hides bar and reset sequence on mouse out
  const resetToMiddleSequence = () => {
    setBarOffset(-1);
    setSequencesOffset(calculateSequenceOffset(sequencesCount / 2));
  };

  useEffect(() => {
    resetToMiddleSequence();
  }, [expanded]);

  return (
    <SequencesWrapper
      className={className}
      onMouseMove={handleMouseMove}
      onMouseOut={resetToMiddleSequence}
      ref={wrapperRef}
    >
      {barOffset >= 0 && (
        <OffsetBar style={{ transform: `translateX(${barOffset}px)` }} />
      )}
      <SequenceHolder
        style={{ backgroundPositionY: `-${sequencesOffset}px` }}
        sequencesSrc={sequencesSrc}
      />
    </SequencesWrapper>
  );
};

export default VideoSequences;
