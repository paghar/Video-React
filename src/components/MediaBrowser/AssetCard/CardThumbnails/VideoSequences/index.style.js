//@flow
import styled from 'styled-components';
import { Secondary } from 'theme/colors';

export const SequencesWrapper = styled.div`
  outline: none;
  display: flex;
  width: 100%;
  height: 100%;
  position: relative;
`;

export const SequenceHolder = styled.div`
  content: ' ';
  width: 100%;
  height: 100%;
  background-image: url(${({ sequencesSrc }) => sequencesSrc});
  background-repeat: no-repeat;
  background-size: 100%;
  z-index: 1;
`;

export const OffsetBar = styled.div`
  content: ' ';
  width: 3px;
  height: 100%;
  position: absolute;
  z-index: 2;
  pointer-events: none;
  background-color: ${Secondary.iii};
`;
