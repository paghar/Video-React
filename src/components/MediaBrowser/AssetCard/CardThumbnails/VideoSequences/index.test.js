//@flow
import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import VideoSequences from '.';
import sequencesImage from 'assets/image/thumb_scrub.jpg';

describe('<AssetCardUnknownThumbnail />', () => {
  it('VideoSequences Thumbnail Snapshot test when there is no mouse hover', () => {
    const tree = renderer
      .create(
        <VideoSequences
          sequencesSrc={sequencesImage}
          sequencesCount={50}
          sequenceSize={{ width: 240, height: 135 }}
          width={250}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('VideoSequences Thumbnail Snapshot test when there is mouse hover on (30,0)', () => {
    const wrapper = mount(
      <VideoSequences
        sequencesSrc={sequencesImage}
        sequencesCount={50}
        sequenceSize={{ width: 240, height: 135 }}
        width={250}
      />
    );
    wrapper.simulate('mousemove', {
      pageX: 30,
    });
    expect(wrapper).toMatchSnapshot();
    wrapper.simulate('mouseout');
    expect(wrapper).toMatchSnapshot();
  });
});
