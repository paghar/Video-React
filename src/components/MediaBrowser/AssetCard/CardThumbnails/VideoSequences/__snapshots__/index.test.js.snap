// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AssetCardUnknownThumbnail /> VideoSequences Thumbnail Snapshot test when there is mouse hover on (30,0) 1`] = `
.c0 {
  outline: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 100%;
  position: relative;
}

.c2 {
  content: ' ';
  width: 100%;
  height: 100%;
  background-image: url(thumb_scrub.jpg);
  background-repeat: no-repeat;
  background-size: 100%;
  z-index: 1;
}

.c1 {
  content: ' ';
  width: 3px;
  height: 100%;
  position: absolute;
  z-index: 2;
  pointer-events: none;
}

<VideoSequences
  sequenceSize={
    Object {
      "height": 135,
      "width": 240,
    }
  }
  sequencesCount={50}
  sequencesSrc="thumb_scrub.jpg"
  width={250}
>
  <styled.div
    onMouseMove={[Function]}
    onMouseOut={[Function]}
  >
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bdVaJa",
            "isStatic": false,
            "lastClassName": "c0",
            "rules": Array [
              "
  outline: none;
  display: flex;
  width: 100%;
  height: 100%;
  position: relative;
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bdVaJa",
          "target": "div",
          "toString": [Function],
          "usesTheme": false,
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={
        Object {
          "current": .c0 {
  outline: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 100%;
  position: relative;
}

.c2 {
  content: ' ';
  width: 100%;
  height: 100%;
  background-image: url(thumb_scrub.jpg);
  background-repeat: no-repeat;
  background-size: 100%;
  z-index: 1;
}

.c1 {
  content: ' ';
  width: 3px;
  height: 100%;
  position: absolute;
  z-index: 2;
  pointer-events: none;
}

<div
            class="c0"
          >
            <div
              class="c1"
              style="transform: translateX(30px);"
            />
            <div
              class="c2"
              style="background-position-y: -NaNpx;"
            />
          </div>,
        }
      }
      onMouseMove={[Function]}
      onMouseOut={[Function]}
    >
      <div
        className="c0"
        onMouseMove={[Function]}
        onMouseOut={[Function]}
      >
        <styled.div
          style={
            Object {
              "transform": "translateX(30px)",
            }
          }
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-htpNat",
                  "isStatic": false,
                  "lastClassName": "c1",
                  "rules": Array [
                    "
  content: ' ';
  width: 3px;
  height: 100%;
  position: absolute;
  z-index: 2;
  pointer-events: none;
  background-color: ",
                    [Function],
                    ";
",
                  ],
                },
                "displayName": "styled.div",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-htpNat",
                "target": "div",
                "toString": [Function],
                "usesTheme": false,
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            style={
              Object {
                "transform": "translateX(30px)",
              }
            }
          >
            <div
              className="c1"
              style={
                Object {
                  "transform": "translateX(30px)",
                }
              }
            />
          </StyledComponent>
        </styled.div>
        <styled.div
          sequencesSrc="thumb_scrub.jpg"
          style={
            Object {
              "backgroundPositionY": "-NaNpx",
            }
          }
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bwzfXH",
                  "isStatic": false,
                  "lastClassName": "c2",
                  "rules": Array [
                    "
  content: ' ';
  width: 100%;
  height: 100%;
  background-image: url(",
                    [Function],
                    ");
  background-repeat: no-repeat;
  background-size: 100%;
  z-index: 1;
",
                  ],
                },
                "displayName": "styled.div",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bwzfXH",
                "target": "div",
                "toString": [Function],
                "usesTheme": false,
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            sequencesSrc="thumb_scrub.jpg"
            style={
              Object {
                "backgroundPositionY": "-NaNpx",
              }
            }
          >
            <div
              className="c2"
              style={
                Object {
                  "backgroundPositionY": "-NaNpx",
                }
              }
            />
          </StyledComponent>
        </styled.div>
      </div>
    </StyledComponent>
  </styled.div>
</VideoSequences>
`;

exports[`<AssetCardUnknownThumbnail /> VideoSequences Thumbnail Snapshot test when there is mouse hover on (30,0) 2`] = `
.c0 {
  outline: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 100%;
  position: relative;
}

.c1 {
  content: ' ';
  width: 100%;
  height: 100%;
  background-image: url(thumb_scrub.jpg);
  background-repeat: no-repeat;
  background-size: 100%;
  z-index: 1;
}

<VideoSequences
  sequenceSize={
    Object {
      "height": 135,
      "width": 240,
    }
  }
  sequencesCount={50}
  sequencesSrc="thumb_scrub.jpg"
  width={250}
>
  <styled.div
    onMouseMove={[Function]}
    onMouseOut={[Function]}
  >
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bdVaJa",
            "isStatic": false,
            "lastClassName": "c0",
            "rules": Array [
              "
  outline: none;
  display: flex;
  width: 100%;
  height: 100%;
  position: relative;
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bdVaJa",
          "target": "div",
          "toString": [Function],
          "usesTheme": false,
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={
        Object {
          "current": <div
            class="c0"
          >
            <div
              class="c1"
              style="background-position-y: -0px;"
            />
          </div>,
        }
      }
      onMouseMove={[Function]}
      onMouseOut={[Function]}
    >
      <div
        className="c0"
        onMouseMove={[Function]}
        onMouseOut={[Function]}
      >
        <styled.div
          sequencesSrc="thumb_scrub.jpg"
          style={
            Object {
              "backgroundPositionY": "-0px",
            }
          }
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-bwzfXH",
                  "isStatic": false,
                  "lastClassName": "c1",
                  "rules": Array [
                    "
  content: ' ';
  width: 100%;
  height: 100%;
  background-image: url(",
                    [Function],
                    ");
  background-repeat: no-repeat;
  background-size: 100%;
  z-index: 1;
",
                  ],
                },
                "displayName": "styled.div",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-bwzfXH",
                "target": "div",
                "toString": [Function],
                "usesTheme": false,
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            sequencesSrc="thumb_scrub.jpg"
            style={
              Object {
                "backgroundPositionY": "-0px",
              }
            }
          >
            <div
              className="c1"
              style={
                Object {
                  "backgroundPositionY": "-0px",
                }
              }
            />
          </StyledComponent>
        </styled.div>
      </div>
    </StyledComponent>
  </styled.div>
</VideoSequences>
`;

exports[`<AssetCardUnknownThumbnail /> VideoSequences Thumbnail Snapshot test when there is no mouse hover 1`] = `
.c0 {
  outline: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 100%;
  position: relative;
}

.c1 {
  content: ' ';
  width: 100%;
  height: 100%;
  background-image: url(thumb_scrub.jpg);
  background-repeat: no-repeat;
  background-size: 100%;
  z-index: 1;
}

<div
  className="c0"
  onMouseMove={[Function]}
  onMouseOut={[Function]}
>
  <div
    className="c1"
    style={
      Object {
        "backgroundPositionY": "-0px",
      }
    }
  />
</div>
`;
